### PR TITLE
[SPARK-43518][SQL] Convert `_LEGACY_ERROR_TEMP_2029` to INTERNAL_ERROR

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to convert _LEGACY_ERROR_TEMP_2029 to INTERNAL_ERROR.

### Why are the changes needed?
1. I found that it can only be triggered it with the parameter value: UP,DOWN,HALF_DOWN,UNNECESSARY, but from a user's perspective, it is impossible (the internal code limits its value to only: HALF_UP,HALF_EVEN,CEILING,FLOOR), so we should convert it to an internal error.
2. The changes improve the error framework.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
1. Update existed UT.
2. Pass GA.

### CODE REVIEW

**Comment by panbingkun** [comment]
friendly cc @MaxGekk

**Review by MaxGekk — APPROVED** [review]
LGTM except of a comment.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala [hunk 1]
```diff
@@ -303,6 +303,17 @@ class DecimalSuite extends SparkFunSuite with PrivateMethodTester with SQLHelper
     }
   }
 
+  test("Not supported rounding mode: UNNECESSARY") {
```
[the remark this answers is not in the document]
Comment (panbingkun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/types/DecimalSuite.scala:306]: haha, let me fix it

**Comment by MaxGekk** [comment]
I have checked the last commit locally.

Merging to master. Thank you, @panbingkun.

**Pull request closed without merging**